src/VisualStudio/Core/Def/Implementation/Snippets/AbstractSnippetExpansionClient.cs

```diff
@@ -393,7 +452,317 @@ public virtual bool TryInsertExpansion(int startPositionInSubjectBuffer, int end
+                {
+                    if (ExpansionSession.GetFieldValue(previousParameter.Name, out var previousValue) == VSConstants.S_OK)
+                    {
+                        _state._arguments = _state._arguments.SetItem(previousParameter.Name, previousValue);
+                    }
+                }
+            }
+
+            var textSpan = new VsTextSpan[1];
+            if (ExpansionSession is null || ExpansionSession.GetSnippetSpan(textSpan) != VSConstants.S_OK)
+            {
+                return;
+            }
+
+            var adjustedTextSpan = textSpan[0];
+            var firstField = _state._symbol?.Parameters.FirstOrDefault()?.Name ?? "placeholder";
+            if (ExpansionSession.GetFieldSpan(firstField, textSpan) != VSConstants.S_OK)
+            {
+                return;
+            }
+
+            adjustedTextSpan.iStartLine = textSpan[0].iStartLine;
+            adjustedTextSpan.iStartIndex = textSpan[0].iStartIndex;
+
+            var lastField = _state._symbol?.Parameters.LastOrDefault()?.Name ?? "placeholder";
+            if (ExpansionSession.GetFieldSpan(lastField, textSpan) != VSConstants.S_OK)
+            {
+                return;
+            }
+
+            adjustedTextSpan.iEndLine = textSpan[0].iEndLine;
+            adjustedTextSpan.iEndIndex = textSpan[0].iEndIndex;
+
+            var snippet = CreateSnippet(symbol.Name, includeMethod: false, symbol.Parameters, cancellationToken);
+            var doc = new DOMDocumentClass();
+            if (doc.loadXML(snippet.ToString(SaveOptions.OmitDuplicateNamespaces)))
+            {
+                // Avoid clearing symbol information when InsertSpecificExpansion ends the current snippet session; the
+                // new session will need the same information to carry argument values forward.
+                _state._preserveSymbols = true;
+
+                Debug.Assert(_state._symbols == symbols);
+                _state._symbol = symbol;
+                var previousArguments = _state._arguments;
+
+                if (expansion.InsertSpecificExpansion(doc, adjustedTextSpan, this, LanguageServiceGuid, pszRelativePath: null, out _state._expansionSession) == VSConstants.S_OK)
+                {
+                    _state._preserveSymbols = false;
+                    Debug.Assert(_state._symbols == symbols);
+                    Debug.Assert(_state._symbol == symbol);
+                    Debug.Assert(_state._arguments == previousArguments);
+
+                    // Even though the closing parenthesis is not part of the updated snippet, make sure the $end$
+                    // position lies after it.
+                    if (ExpansionSession.GetEndSpan(textSpan) == VSConstants.S_OK)
+                    {
+                        textSpan[0].iStartIndex++;
+                        textSpan[0].iEndIndex++;
+                        ExpansionSession.SetEndSpan(textSpan[0]);
+                    }
```

Comment:
don't really understand what's going on here.

Reply:
horribleness is going on 😂

Reply:
I'll figure out a way to document this in code, but the high level version is we are doing strange things to trick the editor into replacing a section of code and starting a snippet session without dismissing the currently-shown Signature Help popup. If the snippet replacement operation alters the `(` or `)` in any way (including overwriting them with the same character), it will dismiss signature help¹. However, we still need <kbd>Enter</kbd> to place the caret outside the `)`. Since there is no snippet XML representation for this concept, we omit `$end$` from the snippet and use the APIs to instruct the editor regarding the logical `$end$` of the snippet even though it's outside the snippet bounds.

¹ This is bad because the editor loses track of which signature is selected. If more than one signature in the list have the same number of parameters, it can make it effectively impossible for a user to scroll through the list because every time it moves to one of the overloads it immediately resets to a different one. The integration tests include a test for this, where a `ToString` call has two overloads that each have one parameter.

Reply:
⏱ updated